docs/source/introduction.md

```diff
@@ -531,6 +531,79 @@ You can also use a dictionary to specify either `dataset_options` or
 >>> write_dataframe(df, '/tmp/test.gpkg', dataset_options={"version": "1.0"}, layer_options={"geometry_name": "the_geom"})
 ```
 
+## Reading from and writing to in-memory datasets
+
+Pyogrio can automatically create and manage in-memory virtual files (`/vsimem`)
+to enable reading and writing from in-memory datasets.
+
+It is possible to read from a dataset stored as bytes:
+
+```python
+from io import BytesIO
+
+# save a GeoJSON to bytes
+geojson = """{
+    "type": "FeatureCollection",
+    "features": [
+        {
+            "type": "Feature",
+            "properties": { },
+            "geometry": { "type": "Point", "coordinates": [1, 1] }
+        }
+    ]
+}"""
+
+geojson_bytes = BytesIO(geojson.encode("UTF-8"))
+
+df = read_dataframe(geojson_bytes)
+```
+
+Note: this may emit a `RuntimeWarning` where the in-memory dataset is detected
+to be a particular format but lacks the expected file extension (e.g., `.gpkg`)
+because the in-memory path automatically created by pyogrio does not include the
+extension.
```

Comment:
Is there a specific reason why the file extension is not added to the filename (to avoid this warning)?

Reply:
because we get only bytes, not a path - so we'd have to inspect the contents of the bytes to try to deduce the file type.  Since GDAL already has the logic in place for that, it seemed better to just leave that to GDAL and not try to look for file type identifiers in the first few bytes and potentially get it wrong.

Reply:
Hmm... true.